boat_simulator/nodes/low_level_control/controller.py

```diff
@@ -0,0 +1,293 @@
+
+        Args:
+            `desired` (Scalar): New desired heading
+            `current` (Scalar): New current heading
+
+        Returns:
+            Scalar: The error between the given headings in radians.
+        """
+        current_bound = bound_to_180(current)
+        desired_bound = bound_to_180(desired)
+        desired_rad = np.deg2rad(desired)
+        current_rad = np.deg2rad(current)
+        error = atan2(sin(desired_rad - current_rad), cos(desired_rad - current_rad))
+
+        if current_bound == desired_bound:
+            error = 0
+
+        return error
+
+    def compute_feedback_angle(self) -> Scalar:
+        """Computes the feedback angle from the given error (in radians)
+        Calculations done using Raye's implementation
+            https://github.com/UBCSailbot/raye-boat-controller/blob/master/python/tack_controller.py
+
+        Returns:
+            Scalar: Difference between current and target rudder angle in degrees.
+        """
+        error = self.compute_error(self.current_heading, self.desired_heading)
+
+        if abs(error) > pi:
+            raise ValueError("heading_error must be between -pi and pi")
+
+        rudder_change = (self.kp * error) / (1 + (self.cp * error))
+        self.running_error = np.rad2deg(rudder_change)
+
+        return np.rad2deg(rudder_change)
+
+    def compute_setpoint(self) -> Scalar:
+        """Calculates the rudder setpoint angle. Only called at the very beginning once
+        to determine the setpoint angle
+
+        Returns:
+            The rudder setpoint angle in degrees
+        """
+        return self.prev_control[0] + self.compute_feedback_angle()
+
+    def update_state(self, speed=2) -> None:  # default speed set to 2 degrees / sec
+        """Updates the rudder angle iteratively towards the target rudder setpoint
+
+        Args:
+            `speed` (Scalar): Speed of rudder angle change in degrees per second
+        """
+        change = speed * self.time_step
+        if self.running_error > change:
+            next_control = self.current_control_ang + change
+            if next_control > self.max_angle_range[0]:
+                next_control = self.max_angle_range[0]
+            elif next_control < self.max_angle_range[1]:
+                next_control = self.max_angle_range[1]
+            else:
+                next_control = next_control
+            self.running_error -= change
+            self.current_control_ang = next_control
+
+        else:
+            next_control = self.current_control_ang + self.running_error
+            if next_control > self.max_angle_range[0]:
+                next_control = self.max_angle_range[0]
+            elif next_control < self.max_angle_range[1]:
+                next_control = self.max_angle_range[1]
+            else:
+                next_control = next_control
+            self.running_error = 0
+            self.current_control_ang = next_control
```

Comment:
This function could be simplified a little more

Reply:
Definitely could be and will be simplified tomorrow